File: Business/src/main/java/name/benjaminabbitt/evented/wwwwwh/url/BusinessLogic/Bookmarks/Bookmark.java
package name.benjaminabbitt.evented.wwwwwh.url.BusinessLogic.Bookmarks;

import name.benjaminabbitt.evented.wwwwwh.url.BusinessLogic.Bookmarks.StateMachine.InvalidState;
import name.benjaminabbitt.evented.bookmarks.Bookmarks;
import name.benjaminabbitt.evented.core.Evented;

import java.util.stream.Stream;

public interface Bookmark {
    Bookmark handle(Bookmarks.BookmarkDeleted bookmarkDeleted) throws InvalidState;

    Stream<Evented.EventPage.Builder> handle(Bookmarks.DeleteBookmark deleteBookmark) throws InvalidState;

    Bookmark handle(Bookmarks.BookmarkCreated bookmarkCreated) throws InvalidState;

    Stream<Evented.EventPage.Builder> handle(Bookmarks.CreateBookmark createBookmark) throws InvalidState;
}

File: BusinessLogic/src/main/kotlin/name/benjaminabbitt/wwwwwh/where/url/businesslogic/statemachine/states/Bookmark.kt
package name.benjaminabbitt.wwwwwh.where.url.businesslogic.statemachine.states

import name.benjaminabbitt.wwwwwh.where.url.projection.sample.bookmarks.Bookmarks.*
import name.benjaminabbitt.wwwwwh.where.url.projection.sample.core.Evented
import name.benjaminabbitt.wwwwwh.where.url.businesslogic.statemachine.InvalidState
import org.slf4j.Logger
import java.util.*
import java.util.stream.Stream

abstract class Bookmark(val id: UUID, protected val logger: Logger) {
    @Throws(InvalidState::class)
    abstract fun handle(bookmarkDeleted: BookmarkDeleted): Bookmark

    @Throws(InvalidState::class)
    abstract fun handle(deleteBookmark: DeleteBookmark): Stream<Evented.EventPage.Builder>

    @Throws(InvalidState::class)
    abstract fun handle(bookmarkCreated: BookmarkCreated): Bookmark

    @Throws(InvalidState::class)
    abstract fun handle(createBookmark: CreateBookmark): Stream<Evented.EventPage.Builder>
}